pkg/virt-controller/watch/snapshot_base.go

```diff
@@ -0,0 +1,517 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	kubevirtv1 "kubevirt.io/client-go/api/v1"
+	vmssv1alpha1 "kubevirt.io/client-go/apis/snapshot/v1alpha1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/controller"
+)
+
+const (
+	volumeSnapshotCRD      = "volumesnapshots.snapshot.storage.k8s.io"
+	volumeSnapshotClassCRD = "volumesnapshotclasses.snapshot.storage.k8s.io"
+)
+
+type informerFunc func(kubecli.KubevirtClient, time.Duration) cache.SharedIndexInformer
+
+var supportedCRDVersions = []string{"v1beta1"}
+
+type dynamicInformer struct {
+	stopCh   chan struct{}
+	informer cache.SharedIndexInformer
+	mutex    sync.Mutex
+
+	informerFunc informerFunc
+}
+
+// SnapshotController is resonsible for snapshotting VMs
+type SnapshotController struct {
+	client kubecli.KubevirtClient
+
+	vmSnapshotQueue        workqueue.RateLimitingInterface
+	vmSnapshotContentQueue workqueue.RateLimitingInterface
+	crdQueue               workqueue.RateLimitingInterface
+
+	vmSnapshotInformer        cache.SharedIndexInformer
+	vmSnapshotContentInformer cache.SharedIndexInformer
+	vmInformer                cache.SharedIndexInformer
+
+	storageClassInformer cache.SharedIndexInformer
+	pvcInformer          cache.SharedIndexInformer
+	crdInformer          cache.SharedIndexInformer
+
+	dynamicInformerMap map[string]*dynamicInformer
+	eventHandlerMap    map[string]cache.ResourceEventHandlerFuncs
+
+	recorder record.EventRecorder
+
+	resyncPeriod time.Duration
+}
+
+// NewSnapshotController creates a new SnapshotController
+func NewSnapshotController(
+	client kubecli.KubevirtClient,
+	vmSnapshotInformer cache.SharedIndexInformer,
+	vmSnapshotContentInformer cache.SharedIndexInformer,
+	vmInformer cache.SharedIndexInformer,
+	storageClassInformer cache.SharedIndexInformer,
+	pvcInformer cache.SharedIndexInformer,
+	crdInformer cache.SharedIndexInformer,
+	recorder record.EventRecorder,
+	resyncPeriod time.Duration,
+) *SnapshotController {
+
+	ctrl := &SnapshotController{
+		client:                    client,
+		vmSnapshotQueue:           workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "snapshot-controller-vmsnapshot"),
+		vmSnapshotContentQueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "snapshot-controller-vmsnapshotcontent"),
+		crdQueue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "snapshot-controller-crd"),
+		vmSnapshotInformer:        vmSnapshotInformer,
+		vmSnapshotContentInformer: vmSnapshotContentInformer,
+		storageClassInformer:      storageClassInformer,
+		pvcInformer:               pvcInformer,
+		vmInformer:                vmInformer,
+		crdInformer:               crdInformer,
+		recorder:                  recorder,
+		resyncPeriod:              resyncPeriod,
+	}
+
+	ctrl.dynamicInformerMap = map[string]*dynamicInformer{
```

Comment:
what happens if someone attempts to create a vmsnapshot when the volume snapshot CRD is not registered yet?

I believe we had a similar issue with datavolumes. our validation webhook should be rejecting vmi/vm with datavolume if the datavolume crd isn't registered.

Reply:
> what happens if someone attempts to create a vmsnapshot when the volume snapshot CRD is not registered yet?

The snapshot will still get created.  We want to be able to snapshot VMs that don't reference PVCs.  The question is how to handle VMs with PVCs that cannot be snapshotted.  I think it is still okay to proceed with the snapshot.  Maybe the user only cares about backing up/restoring their VM config.  But maybe you/others don't agree?  If so, we can of course do something with a webhook.